### ISSUE TITLE
Unable to Build with New Installation (firmware v2)

### ISSUE THREAD

**Issue by jdlien** [jdlien/Kaleidoscope-LEDEffect-FunctionalColor#7]
Using the latest version of the Arduino IDE (1.8.8) from arduino.cc and and getting the latest version of Keleidoscope, I am unable to build an .ino with a FunctionalColors object defined.

I've been poking at this for a while and can't figure out why the compiler is complaining about undefined references. The location it mentions is definitely the latest version of FunctionalColor.

I'm at a loss here... In my test .ino I just took the default included one from Kaleidoscope and added these lines after the initial includes:

```c++
#include <Kaleidoscope-LEDEffect-FunctionalColor.h>
using namespace kaleidoscope::LEDFunctionalColor;
FCPlugin funColor;
```

The complete compiler output follows, in case it reveals something I didn't already mention.

```
WARNING: Category 'Debug' in library Kaleidoscope-Hardware-Virtual is not valid. Setting to 'Uncategorized'
libraries/Kaleidoscope-LEDEffect-FunctionalColor/Kaleidoscope-LEDEffect-FunctionalColor.a(Kaleidoscope-LEDEffect-FunctionalColor.cpp.o): In function `cRGB kaleidoscope::LEDFunctionalColor::groupColorLookup<kaleidoscope::LEDFunctionalColor::colorMapDefault>(kaleidoscope::Key const&, bool&)':
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::control'
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::control'
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::control'
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::gui'
Multiple libraries were found for "HID.h"
 Used: /Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/HID
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::gui'
 Not used: /Applications/Arduino.app/Contents/Java/hardware/arduino/avr/libraries/HID
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::gui'
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::shift'
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::shift'
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::shift'
collect2: error: ld returned 1 exit status
exit status 1
Error compiling for board Keyboardio Model 01.
```

To answer questions posed in https://github.com/jdlien/Kaleidoscope-LEDEffect-FunctionalColor/pull/5#issuecomment-466755746, I am building this with the IDE on macOS Mojave. As long as I comment out "FCPlugin funColor" this will compile. Interestingly, it also compiles if I change it to "FCPlugin funColor", although giving a numerical argument (which would normally set brightness) doesn't work.

**Comment by noseglasses** [comment]
I tried the exact same as described above with both Arduino 1.8.5 and 1.8.8 on Ubuntu 18.04 and it builds and links flawlessly. As I have no access to an Apple so I cannot reproduce your build problems.

My Arduino 1.8.8's version of avr-g++ is 5.4. Could you check yours so that we can compare. The source of the problem might as well be the linker which would explain why it behaves differently on different systems. I have the suspicion that the `inline constexpr cRGB dim(const cRGB &color, byte brightness, byte range=255)` method in `Kaleidoscope-LEDEffect-FunctionalColor/src/colors.h` is the culprit. 

Could you try and convert its signature to `inline constexpr cRGB dim(cRGB color, byte brightness, byte range=255)` and see if this fixes your problems. 

For any strange reasons the compiler appears to generate instances of variables that are actually supposed to be compile time constants. Taking a reference through the `dim(...)` method might be a reason but this should only happen, if the reference is taken in a non-constexpr context which is not the case here.

All this does not explain, why the plugin used to work for you and then after the transition to v2 ceased to work.

This is very strange.

**Comment by ian-kelling** [comment]
I've got the same problem. Same version of avr-g++. Changing the line in colors.h did not help.

**Comment by noseglasses** [comment]
@ian-kelling, thanks for trying this.

Which version of Kaleidoscope-Bundle-Keyboardio do you use? Recently (~10 days ago) I submitted a fix for a linker problem that might as well be related. If you are not yet on the latest version of the bundle, could you please give it a try?

**Comment by jdlien** [comment]
@noseglasses I'm having another crack at seeing if I can get this to compile and refreshed my environment, setting it up again (still on macOS). Same issues as before.

Removing the & from colors.h line 180 didn't make a difference.
```C++
inline constexpr cRGB dim(const cRGB &color, byte brightness, byte range=255) {
```
Presuming I just did git clone of the latest Kaleidoscope-Bundle-Keyboardio, your linker fix would be included, right? Or is it in a separate branch?

**Comment by jdlien** [comment]
I've been experimenting and switching things around for a few hours and don't feel much closer to solving this issue. But one thing I tried was removing all colors from every colorMap struct. When I do this, it changes the error I get considerably:

```
libraries/Kaleidoscope-LEDEffect-FunctionalColor/Kaleidoscope-LEDEffect-FunctionalColor.a(Kaleidoscope-LEDEffect-FunctionalColor.cpp.o): In function `cRGB kaleidoscope::plugin::groupColorLookup<kaleidoscope::plugin::colorMapDefault>(kaleidoscope::Key const&, bool&)':
/Users/jdlien/Documents/Arduino/hardware/keyboardio/avr/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:18: undefined reference to `kaleidoscope::plugin::colorMap::defaultColor'
/Users/jdlien/Documents/Arduino/hardware/keyboardio/avr/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:18: undefined reference to `kaleidoscope::plugin::colorMap::defaultColor'
/Users/jdlien/Documents/Arduino/hardware/keyboardio/avr/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:18: undefined reference to `kaleidoscope::plugin::colorMap::defaultColor'
```

Is the program not seeing the definitions of items in colorMap for some reason? Do I have to change the way those are all defined?

It's interesting that these errors always come in groups of three. I'm not sure why that would be.

**Comment by jdlien** [comment]
Okay, I think I have almost figured this out. It is related to the subject of this StackOverflow article.
[How to initialize private static members in C++?](https://stackoverflow.com/questions/185844/how-to-initialize-private-static-members-in-c)

It seems that (if my understanding is correct, and it probably isn't) the linker errors because each file including the header has its own definition of the static members.

By removing the constexpr keyword, and not defining the values of the structs in the .h file, and instead moving them to the cpp file, this actually does compile and functions.

However, this is tedious, messy, and, I suspect, considerably less memory efficient.

The resulting code in the .h looks something like:
```C++
struct colorMapDefault : public colorMap {
  static cRGB shift;
  static cRGB control;
  static cRGB gui;
  //etc...
};
```

And in the .cpp it looks like
```C++
  cRGB colorMapDefault::shift = dim(darkseagreen, 190);
  cRGB colorMapDefault::control = dim(lightskyblue, 190);
  cRGB colorMapDefault::gui = dim(pink, 190);
  //etc...
```

@noseglasses, what do you think the best way forward would be?

It looks like I have figured out a way to make this work, at the cost of some increased code complexity and probably memory use. I'm not sure how much difference the constexpr actually makes, or why it suddenly stopped working the way it used to, or the way it seems to still work fine for you. I'm hoping that there's still some way to make this function with constexpr to get the memory savings from evaluating everything at compile time.

**Comment by jdlien** [comment]
Alright, I think I have a pretty workable solution, although it's not as elegant as I'd like.
I was informed by this discussion: https://stackoverflow.com/questions/8016780/undefined-reference-to-static-constexpr-char#8016853

If I simply leave the existing .h code alone but add a (seemingly pointless) redeclaration of all the variables into the cpp file, this seems to work.
```C++
  constexpr cRGB colorMap::defaultColor;
  constexpr cRGB colorMap::shift;
  constexpr cRGB colorMap::control;
  //etc...
```

I'll implement this fix after I get some sleep and hopefully this will finally be usable for people in my position again!

**Comment by noseglasses** [comment]
> Presuming I just did git clone of the latest Kaleidoscope-Bundle-Keyboardio, your linker fix would be included, right? Or is it in a separate branch?

Right.

> Alright, I think I have a pretty workable solution, although it's not as elegant as I'd like.
I was informed by this discussion: https://stackoverflow.com/questions/8016780/undefined-reference-to-static-constexpr-char#8016853

The weird thing about all this is that the workaround seems not to be necessary when build on Linux. That's something that definitely should not happen when cross building for the same target platform on different host systems.

I suspect that on the macOS build the color variables are ODR-used somewhere. ODR-use can e.g. mean that a reference is taken of such variables. If so, the compiler has to instantiate the variables and thereby ignores them being declared `constexpr`. This totally makes sense as a pointer or reference points to a memory location and `constexpr` variables normally don't come with instances and, thus, are not present in memory.

Your seemingly pointles re-declarations are actually __definitions__ rather than __declarations__. Only those enable the compiler to allocate static memory for the variables.

What is weird about all this is that I cannot see any place where this happens and it appears only to happen on non-Linux systems.

We could compare the binaries from macOS and Linux to find the differences. Could you upload a minimal sketch and firmware binary that includes your workaround and provide details about the exact bundle and plugin version you used to build it. I could then build the same on Linux and see if both binaries differ. Also I could compare the macOS-version with workaround with the Linux version without.

**Comment by jdlien** [comment]
Hi @noseglasses, I've created a [new branch called UndefinedReferenceTest](https://github.com/jdlien/Kaleidoscope-LEDEffect-FunctionalColor/tree/UndefinedReferenceTest).

Something I discovered was that simply adding the redefinitions wasn't enough to get it working. In my previous bludgeoning of this code, I had also changed namespaces around (chiefly to move everything into the typical kaleidoscope::plugin namespace, and I also changed FCPlugin to FunctionalColor to be consistent with naming of other plugins. Let me know if you think I've made a grave error here, but for some reason this seems to help make things work.

I have created an example sketch using the default sketch with everything nonessential removed. You can find that in the examples folder in fc_test_minimal if you want to try compiling it.

I have included my last build in the examples folder as well, in arduino_build_503364.  This was built using the latest version of the kaleidoscope bundle via git clone late last night.

Let me know if you can provide any insight into why our compilers seem to behave so differently. I'm interested in discovering if you are able to compile after commenting out the redefinitions in FunctionalColor.cpp and if so, why that might be.

Thanks again for your help!

**UPDATE:**
In retrospect, changing the namespaces of things in FunctionalColors made a huge mess and make it much more difficult to work with. I'm not against the idea of changing it in some capacity, but I probably don't want to do it the way I did here, because it made working with colors and functions in FunctionalColors a pain. I'd like to better understand why this didn't work until I changed namespace.

**Comment by noseglasses** [comment]
There seems to be a strange consistency in the handling of constexpr static class variables that cause such linker errors (see https://stackoverflow.com/a/28446388 and https://stackoverflow.com/a/43621912). This inconsistency appears to be fixed in later versions of the C++ standard.

As Kaleidoscope is currently using C++-11, I implemented a workaround that replaces all constexpr static color class variables with functions. This has the great advantage that users do not need to define those variables explicitly as it would otherwise be necessary for the example sketches.

By force inlining the color functions and function `dim(...)` the firmware size for the simple example sketch even decreases by 58 bytes (see 
[fc_diff.pdf](https://github.com/jdlien/Kaleidoscope-LEDEffect-FunctionalColor/files/3369086/fc_diff.pdf)).


The namespace problems have also been fixed by introducing a namespace `FunctionalColorStuff` (@jdlien, feel free to rename it, I didn't find anything better).

Here's the PR: https://github.com/jdlien/Kaleidoscope-LEDEffect-FunctionalColor/pull/8

**Comment by jdlien** [comment]
Huh, this is a pretty interesting approach. I probably won't be able to spend much time on this for a bit, however.

There is something I've been thinking about that you mentioned in the past, however. Is there any chance of getting colors.h and/or keygroups.h integrated natively into Kaleidoscope? I built these with the notion that they might be useful to other people, for handy color name shortcuts and the ability to categorize keys, and you seemed to agree.

**Comment by noseglasses** [comment]
> There  is something I've been thinking about that you mentioned in the past,  however. Is there any chance of getting colors.h and/or keygroups.h  integrated natively into Kaleidoscope? I built these with the notion  that they might be useful to other people, for handy color name  shortcuts and the ability to categorize keys, and you seemed to agree.

That's definitely something that might be valuable for other people writing plugins or even within the firmware core. Just submit a PR and see what happens.

**Comment by jdlien** [comment]
Sorry for being a bit off topic in the context of this issue, but...
If I did want to submit a PR, should I make any changes first? In particular I wonder if I should change the namespace/scope of it. Is kaleidoscope::plugin appropriate?

**Comment by noseglasses** [comment]
I guess algernon already answered this in the community forum.

**Comment by noseglasses** [comment]
This issue can probably be closed after https://github.com/jdlien/Kaleidoscope-LEDEffect-FunctionalColor/pull/8 has been merged.

**Comment by jdlien** [comment]
Hi @noseglasses, I've finally had a chance to put a few hours into looking at this. There are some major issues still.

I've pushed a minorly changed version to master (with a couple naming changes FunctionalColorStuff became LEDFunctionalColors for consistency with the old version). I have created a new fc_example.ino that works with the new version of the firmware (except for colorMaps, as detailed below).

For some reason when I try adding my FunctionalColors example into the default firmware, it seems to run the keyboard out of memory. I'm not sure what changed that is causing such high memory usage.

Also, I can't use custom defined themes anymore. When I create colorMap structs I get all manner of errors. I see that you substantially changed the way this worked by making each property a function, but even if I changed it to this format in the firmware, I get quite a lot of errors as below.

```
error: 'baseColor' declared as function returning a function
 #define FC_MAP_COLOR(NAME, ...) __attribute__((always_inline)) static constexpr cRGB NAME() { return __VA_ARGS__; }
                                                                                           ^
/Users/jdlien/kaleidoscope/Kaleidoscope-Bundle-Keyboardio/avr/libraries/Kaleidoscope-LEDEffect-FunctionalColor/examples/fc_example/fc_example.ino:413:3: note: in expansion of macro 'FC_MAP_COLOR'
   FC_MAP_COLOR(FC_REF_MAP_COLOR(baseColor), green)
   ^
```

Any idea what's up with the colorMaps in fc_example.ino?

**Comment by noseglasses** [comment]
I submitted a PR that fixes the example: https://github.com/jdlien/Kaleidoscope-LEDEffect-FunctionalColor/pull/10

This PR also fixes build warnings and enables testing the example sketch with both ways of color specifications (see below).

With respect to the size of the sketch I have no other explanation than that the example already used to be short below the PROGMEM limit and some other part of the firmware grew a bit in the meantime. 
I intented to check the firmware size of the example from say a year ago but I failed to build it as it requires to find a suitable match of plugin commit version and firmware bundle. After some attempts I gave up.

Fortunately it is easy (for testing purposes) to go back to the old way of using static constexpr color values by simply commenting the following line (considering the above PR). 
https://github.com/jdlien/Kaleidoscope-LEDEffect-FunctionalColor/blob/e65aeaddcd778ae7c898495ed1ca66bfa1885efd/src/Kaleidoscope-LEDEffect-FunctionalColor.h#L15

By comparing builds of the example firmware with the old and the new style of color definitions, one can see that the function-based version requires less PROGMEM space but still exceeds the limit.

Final lines of build output:

With `#define FC_HAVE_COLOR_FUNCTIONS` commented (with traditional static constexpr color variables).
```
Sketch uses 29866 bytes (104%) of program storage space. Maximum is 28672 bytes.
Global variables use 1547 bytes (60%) of dynamic memory, leaving 1013 bytes for local variables. Maximum is 2560 bytes.
Sketch too big; ...
```

With the new way of specifying colors.
```
Sketch uses 29308 bytes (102%) of program storage space. Maximum is 28672 bytes.
Global variables use 1475 bytes (57%) of dynamic memory, leaving 1085 bytes for local variables. Maximum is 2560 bytes.
Sketch too big; ...
```

The only way to fix the example is to remove some other plugins or to reduce the amount of functional colors stuff.
